README.markdown
# jQuery MultiSelect Plugin

This plugin turns an ordinary HTML select control into elegant drop down list of checkboxes, stylable with ThemeRoller.  
Based off of Cory S.N. LaViska's implementation, A Beautiful Site (http://www.abeautifulsite.net) 2009.

Demos and documentation @ [http://www.erichynds.com/jquery/jquery-multiselect-plugin-with-themeroller-support/](http://www.erichynds.com/jquery/jquery-multiselect-plugin-with-themeroller-support/)  

![Example](http://www.erichynds.com/examples/jquery-multiselect/screenshot.gif)


CHANGELOG
jQuery MultiSelect Changelog

Version 0.2 - 02/02/2010
==============================================
- Modified: the initial call to updateSelected() does not need to be in an each() loop.
- Modified: refactored $select bindings to use one bind() call, passing an object literal.
- Fixed: removed an unnecessary, commented-out line.
- Added: the selectedText can now take two pound-sign placeholders.  The second is replaced with the total number of checkboxes.

Version 0.3 - 02/04/2010
==============================================
- Added: ability to toggle optgroup check state by clicking on the header.
- Added: onCheckAll, onUncheckAll, and onOptgroupToggle callbacks.
- Added: the selectedList now takes a number denoting the maximum number of items to display as a list until deferring to selectedText.
- Added: the widget's default width now inherits the width of the original select element.
- Added: mixWidth parameter to force a minimum width.
- Added: link to documentation in comments.
- Fixed: global variable "width".
- Modified: standardized using single quotes.
- Modified: refactored removeAttr("checked") to attr("checked", "").

Version 0.4 - 02/15/2010
==============================================
- Added: the down arrow & space toggles open/close, and up arrow closes when focused on the input.
- Added: support for disabled="disabled" options.
- Added: the checkbox inputs now have the "title" attribute.
- Fixed: traversing the menu now skips over disabled options & optgroup headers.
- Fixed: removed the call to focus() when traversing the menu if there isn't a next/prev li.
- Fixed: setting the showHeader option to false no longer breaks the scrollbar or keyboard traversal scrollTop setting.
- Fixed: added relative positioning & z-index to .ui-multiselect-options (thanks David).
- Fixed: the main input's value is no longer cleared when a form's reset button is clicked.
- Fixed: the "title" attribute of the main input updates correctly when the last checkbox is unchecked.
- Modified: added a custom "toggleChecked" event to $options.
- Modified: the minWidth parameter is now 200 by default (from "auto").
- Modified: exposed default options by moving them to $.fn.multiSelect.defaults.

Version 0.4.1 - 02/15/2010
==============================================
- Fixed: graceful handling of form resets with pre-selected options, cross browser.

Version 0.5 - 03/13/2010
==============================================
- Added: support for disabled selects (entire widget).
- Added: the widget now inherits the original ID of the select box.
- Added: "state" parameter added to configure the default open or closed state.
- Modified: selectedText can now take a function with 3 args: number of inputs checked, number of inputs total, an array of the checked inputs.
- Modified: slight refactoring (moved updateSelected et al.).
- Modified: noneSelected parameter renamed to noneSelectedText.
- Modified: positioning now determined by offset() instead of position().
- Modified: switched to "this" from "$(this)" where possible.
- Fixed: No need to pass the "px" unit to width().
- Fixed: The icons in the "check all" and "uncheck all" header links were not toggling correctly.
- Fixed: removed some unused vars.
- Fixed: better positioning of the header links.
- Fixed: border issue on the text input when inside a div.ui-widget-content.
- Fixed: options menu now has the ui-corner-all class.  Previously all corners except top left were rounded... for some reason.
- Fixed: use position() instead of offset() when inside a .ui-widget-content class.  offset() won't position the dropdown correctly.





jquery.multiSelect.js
/*
 * jQuery multiSelect 0.1
 * by Eric Hynds, 2010
 *
 * Inspired by Cory S.N. LaViska's implementation, A Beautiful Site (http://abeautifulsite.net/) 2009
 *
 * Licensing & Terms of Use
 * 
 * Dual licensed under the MIT and GPL licenses:
 *   http://www.opensource.org/licenses/mit-license.php
 *   http://www.gnu.org/licenses/gpl.html
 *
*/

(function($){
	
	$.fn.extend({
		multiSelect: function(opts){
			opts = $.extend({}, MultiSelect.defaults, opts);

			return this.each(function(){
				new MultiSelect(this, opts);
			});
		}
	});
	
	var MultiSelect = function(select,o){
		var $select = $original = $(select), $options, $labels, html = '', optgroups = [];
		
		//html += '<a class="multiSelect ui-state-default ui-corner-all"><input readonly="readonly" type="text" value="" /><img src="arrow.gif" class="multiSelect-arrow" alt="" /></a>';
		html += '<a class="ui-multiselect ui-state-default ui-corner-all"><input readonly="readonly" type="text" value="" /><span class="ui-icon ui-icon-triangle-1-s"></span></a>';
		html += '<div class="ui-multiselect-options' + (o.shadow ? ' ui-multiselect-shadow' : '') + ' ui-widget ui-widget-content ui-corner-bl ui-corner-br ui-corner-tr">';
	
		if(o.showHeader){
			html += '<div class="ui-widget-header ui-helper-clearfix ui-corner-all ui-multiselect-header">';
			html += '<ul class="ui-helper-reset">';
			html += '<li><a class="ui-multiselect-all" href=""><span class="ui-icon ui-icon-check"></span>' + o.checkAllText + '</a></li>';
			html += '<li><a class="ui-multiselect-none" href=""><span class="ui-icon ui-icon-closethick"></span>' + o.unCheckAllText + '</a></li>';
			html += '<li class="ui-multiselect-close"><a href="" class="ui-multiselect-close ui-icon ui-icon-circle-close"></a></li>';
			html += '</ul>';
			html += '</div>';
		};
		
		// build options
		html += '<ul class="ui-multiselect-checkboxes ui-helper-reset">';
		$select.find('option').each(function(i){
			var $this = $(this), value = $this.val(), len = value.length, $parent = $this.parent();
			
			if($parent.is("optgroup")){
				var label = $parent.attr("label");
				
				if($.inArray(label, optgroups) === -1){
					html += '<li class="ui-multiselect-optgroup-label"><span>' + label + '</span></li>';
					optgroups.push(label);
				}
			}
			
			if(len > 0){
				html += $this.parent().is("optgroup") ? '<li class="multiSelect-optgroup">' : '<li>';
				html += '<label class="ui-corner-all"><input type="checkbox" name="' + $select.attr('name') + '" value="' + value + '"';
				if($this.is(':selected')) html += ' checked="checked"';
				html += ' />' + $this.html() + '</label></li>';
			};
		});
		html += '</ul></div>';
		
		// cache queries
		$select  = $select.after(html).next('a.ui-multiselect');
		$options = $select.next('div.ui-multiselect-options');
		$header  = $options.find('div.ui-multiselect-header');
		$labels  = $options.find("label");
		
		// the select box events
		$select
		.bind('mouseover mouseout', function(e){
			$(this)[ (e.type === 'mouseover') ? 'addClass' : 'removeClass' ]('ui-state-hover');
		})
		.bind('focus blur', function(e){
			$(this)[ (e.type === 'focus') ? 'addClass' : 'removeClass' ]('ui-state-focus');
		})
		.bind('click', function(e){
			$options.trigger( $options.is(':hidden') ? 'open' : 'close' );
		})
		.bind('keypress', function(e){
			switch(e.keyCode){
				case 27: // esc
					$options.trigger('close');
					break;
			};
		})
		.each(function(){
			// update the number of selected elements when the page initally loads
			updateSelected();
		});
		
		
		// header links
		if(o.showHeader){
			$header.find("a").click(function(e){
				var $target = $(e.target);
			
				// close link
				if($target.hasClass("ui-multiselect-close")){
					$options.trigger('close');
			
				// check all / uncheck all
				} else {
					$options.trigger('toggleChecked', [($target.hasClass("ui-multiselect-all") ? true : false)]);
				}
			
				e.preventDefault();
			});
		};
		
		// bind custom events to the options div
		$options.bind({
			'close': function(e, others){
				others = others || false;
			
				// hides all other options but the one clicked
				if(others === true){
					$('div.ui-multiselect-options')
					.filter(':visible')
					.fadeOut(o.fadeSpeed)
					.prev("a.ui-multiselect")
					.removeClass('ui-state-active')
					.trigger('mouseout');
					
				// hides the clicked options
				} else {
					$select.removeClass('ui-state-active').trigger('mouseout');
					$options.fadeOut(o.fadeSpeed);
				};
			},
			'open': function(e){
				var offset = $select.position(), $container = $options.find("ul:eq(1)"), timer, listHeight = 0, top;
				
				// calling select is active
				$select.addClass('ui-state-active');
				
				// hide all other options
				$options.trigger("close", [true]);
				
				// calculate positioning
				if(o.position === 'middle'){
					top = ( offset.top+($select.height()/2)-($options.outerHeight()/2) );
				} else if(o.position === 'top'){
					top = (offset.top-$options.outerHeight());
				} else {
					top = (offset.top + $select.outerHeight());
				};
				
				// calculate the width
				width = $select.width()-parseInt($options.css('padding-left'))-parseInt($options.css('padding-right'));
				
				// select the first option
				$labels.filter("label:first").trigger("mouseenter");
				
				// show the options div + position it
				$options
				.css({ 
					position: 'absolute',
					top: top+'px',
					left: offset.left+'px',
					width: width+'px'
				})
				.show();
				
				// set the scroll of the checkbox container
				$container.scrollTop(0);
				
				// set the height of the checkbox container
				if(o.maxHeight){
					$container.css("height", o.maxHeight );
				};
				
				o.onOpen.call($options[0]);
			},
			'traverse': function(e, start, keycode){
				var $start = $(start), moveToLast = (keycode === 38 || keycode === 37) ? true : false; 
				
				// attempt to move to the next label
				var $next = $start.parent()[ moveToLast ? 'prev' : 'next' ]('li').find('label').trigger('mouseenter');
				
				// if at the first/last element
				if(!$next.length){
					var $container = $options.find("ul:eq(1)");
					
					// move to the first/last
					$options.find('label')[ moveToLast ? 'last' : 'first' ]().trigger('mouseover');
					
					// set scroll position
					$container.scrollTop( moveToLast ? $container.height() : 0 );
				};
			},
			'toggleChecked': function(e, flag){
				var $inputs = $labels.find("input");
				if(flag){
					$inputs.attr("checked","checked"); 
				} else {
					$inputs.removeAttr("checked");
				}
				
				updateSelected();
			}
		});
		
		
		// labels
		$labels
		.bind('mouseenter', function(e){
			$labels.removeClass('ui-state-hover');
			$(this).addClass('ui-state-hover').find("input").focus();
		})
		.bind('click', function(e,caller){
			// if the label was clicked, trigger the click event on the checkbox.  IE6 fix
			e.preventDefault();
			$(this).find("input").trigger("click", [true]);
		})
		.bind('keyup', function(e){
			switch(e.keyCode){
				case 9: // tab
					$options.trigger('close');
					$select.next(":input").focus();
					break;

				case 27: // esc
					$options.trigger('close');
					break;
			
				case 38: // up
				case 40: // down
				case 37: // left
				case 39: // right
					$options.trigger('traverse', [this, e.keyCode]);
					break;
				
				case 13: // enter
					e.preventDefault();
					$(this).click();
					break;
			};
		})
		.find('input')
		.bind('click', function(e, label){
			var $this = $(this);
			label = label || false;
			
			// stop this click from bubbling up to the label
			e.stopPropagation();
			
			// if the click originated from the label, stop the click event and manually toggle the checked state
			if(label){
				e.preventDefault();
				if($this.is(":checked")){
					$this.removeAttr("checked");
				} else {
					$this.attr("checked","checked");
				}
			}
			
			o.onCheck.call( $this[0] );
			updateSelected();
		});
		// apply bgiframe if available
		if($.fn.bgiframe){
			$options.bgiframe();		};
		
		// remove the original input element
		$original.remove();

		function updateSelected(){
			var $input = $select.find("input"),
			    $inputs = $labels.find('input'),
			    $checked = $inputs.filter(":checked"),
			    value = '',
			    numChecked = $checked.length;
			
			if(numChecked === 0){
				$input.val( o.noneSelected );
			} else {
			
				// list items?
				if(o.selectedList){
					$checked.each(function(){
						var text = $(this).parent().text();
						value = (value.length) ? (value += ', ' + text) : text;
					});
				} else {
					value = o.selectedText.replace('#', numChecked);
				};
				
				$input.val( value ).attr("title", value);
			};
		};

		return $options;
	};
	
	// close each select when clicking on any other element/anywhere else on the page
	$(document).bind("click", function(e){
		var $target = $(e.target);

		if(!$target.closest("div.ui-multiselect-options").length && !$target.parent().hasClass("ui-multiselect")){
			$("div.ui-multiselect-options").trigger("close", [true]);
		};
	});
	
	// default options
	MultiSelect.defaults = {
		showHeader: true,
		maxHeight: 175, /* max height of the checkbox container (scroll) in pixels */
		checkAllText: 'Check all',
		unCheckAllText: 'Uncheck all',
		noneSelected: 'Select options',
		selectedList: false,
		selectedText: '# selected',
		position: 'bottom', /* top|middle|bottom */
		shadow: false,
		fadeSpeed: 200,
		onCheck: function(){},
		onOpen: function(){}
	};

})(jQuery);


src/jquery.multiselect.js
/*
 * jQuery MultiSelect Plugin 0.5
 * Copyright (c) 2010 Eric Hynds
 *
 * http://www.erichynds.com/jquery/jquery-multiselect-plugin-with-themeroller-support/
 * Inspired by Cory S.N. LaViska's implementation, A Beautiful Site (http://abeautifulsite.net/) 2009
 *
 * Dual licensed under the MIT and GPL licenses:
 *   http://www.opensource.org/licenses/mit-license.php
 *   http://www.gnu.org/licenses/gpl.html
*/
(function($){
	
	$.fn.multiSelect = function(opts){
		opts = $.extend({}, $.fn.multiSelect.defaults, opts);

		return this.each(function(){
			 return new MultiSelect(this, opts);
		});
	};
	
	var MultiSelect = function(select,o){
		var $select = $original = $(select), $options, $header, $labels, html = [], optgroups = [], isDisabled = $select.is(':disabled');
		
		html.push('<a id="'+ select.id +'" class="ui-multiselect ui-widget ui-state-default ui-corner-all' + (isDisabled || o.disabled ? ' ui-state-disabled' : '') + '">');
		html.push('<input readonly="readonly" type="text" class="ui-state-default" value="'+ o.noneSelectedText +'" /><span class="ui-icon ui-icon-triangle-1-s"></span></a>');
		html.push('<div class="ui-multiselect-options' + (o.shadow ? ' ui-multiselect-shadow' : '') + ' ui-widget ui-widget-content ui-corner-all">');
	
		if(o.showHeader){
			html.push('<div class="ui-widget-header ui-helper-clearfix ui-corner-all ui-multiselect-header">');
			html.push('<ul class="ui-helper-reset">');
			html.push('<li><a class="ui-multiselect-all" href=""><span class="ui-icon ui-icon-check"></span>' + o.checkAllText + '</a></li>');
			html.push('<li><a class="ui-multiselect-none" href=""><span class="ui-icon ui-icon-closethick"></span>' + o.unCheckAllText + '</a></li>');
			html.push('<li class="ui-multiselect-close"><a href="" class="ui-multiselect-close ui-icon ui-icon-circle-close"></a></li>');
			html.push('</ul>');
			html.push('</div>');
		}
		
		// build options
		html.push('<ul class="ui-multiselect-checkboxes ui-helper-reset">');
		
		/* 
			If this select is disabled, remove the actual disabled attribute and let themeroller's .ui-state-disabled class handle it.
			This is a workaround for jQuery bug #6211 where options in webkit inherit the select's disabled property.  This
			won't achieve the same level of 'disabled' behavior (the checkboxes will still be present in form submission), 
			but it at least gives you a way to emulate the UI. 
		*/
		if(isDisabled){
			$select.removeAttr("disabled");
		}
		
		$select.find('option').each(function(){
			var $this = $(this), title = $this.html(), value = this.value, len = value.length, $parent = $this.parent(), hasOptGroup = $parent.is('optgroup'), isDisabled = $this.is(':disabled'), labelClasses = ['ui-corner-all'], liClasses = [];
			
			if(hasOptGroup){
				var label = $parent.attr('label');
				
				if($.inArray(label,optgroups) === -1){
					html.push('<li class="ui-multiselect-optgroup-label"><a href="#">' + label + '</a></li>');
					optgroups.push(label);
				}
			}
		
			if(len > 0){
				if(isDisabled){
					labelClasses.push('ui-state-disabled');
					liClasses.push('ui-multiselect-disabled');
				}
				
				html.push('<li class="' + liClasses.join(' ') + '">');
				html.push('<label class="' + labelClasses.join(' ') + '"><input type="checkbox" name="' + select.name + '" value="' + value + '" title="' + title + '"');
				if($this.is(':selected')){
					html.push(' checked="checked"');
				}
				if(isDisabled){
					html.push(' disabled="disabled"');
				}
				html.push(' />' + title + '</label></li>');
			}
		});
		html.push('</ul></div>');

		// append the new control to the DOM; cache queries
		$select  = $select.after( html.join('') ).next('a.ui-multiselect');
		$options = $select.next('div.ui-multiselect-options');
		$header  = $options.find('div.ui-multiselect-header');
		$labels  = $options.find('label').not('.ui-state-disabled');
		
		// calculate widths
		var iconWidth = $select.find('span.ui-icon').outerWidth(), inputWidth = $original.outerWidth(), totalWidth = inputWidth+iconWidth;
		if( /\d/.test(o.minWidth) && totalWidth < o.minWidth){
			inputWidth = o.minWidth-iconWidth;
			totalWidth = o.minWidth;
		}
		
		// set widths
		$select.width(totalWidth).find('input').width(inputWidth);
		
		// build header links
		if(o.showHeader){
			$header.find('a').click(function(e){
				var $this = $(this);
			
				// close link
				if($this.hasClass('ui-multiselect-close')){
					$options.trigger('close');
			
				// check all / uncheck all
				} else {
					var checkAll = $this.hasClass('ui-multiselect-all');
					$options.trigger('toggleChecked', [(checkAll ? true : false)]);
					o[ checkAll ? 'onCheckAll' : 'onUncheckAll']['call'](this);
				}
			
				e.preventDefault();
			});
		}
		
		var updateSelected = function(){
			var $inputs = $labels.find('input'),
				$checked = $inputs.filter(':checked'),
				value = '',
				numChecked = $checked.length;
			
			if(numChecked === 0){
				value = o.noneSelectedText;
			} else {
				if($.isFunction(o.selectedText)){
					value = o.selectedText.call(this, numChecked, $inputs.length, $checked.get());
				} else if( /\d/.test(o.selectedList) && o.selectedList > 0 && numChecked <= o.selectedList){
					value = $checked.map(function(){ return this.title; }).get().join(', ');
				} else {
					value = o.selectedText.replace('#', numChecked).replace('#', $inputs.length);
				}
			}
			
			$select.find('input').val(value).attr('title', value);
			return value;
		};
		
		// the select box events
		$select.bind({
			click: function(){
				$options.trigger('toggle');
			},
			keypress: function(e){
				switch(e.keyCode){
					case 27: // esc
					case 38: // up
						$options.trigger('close');
						break;
					case 40: // down
					case 0: // space
						$options.trigger('toggle');
						break;
				}
			},
			mouseenter: function(){
				if(!$select.hasClass('ui-state-disabled')){
					$(this).addClass('ui-state-hover');
				}
			},
			mouseleave: function(){
				$(this).removeClass('ui-state-hover');
			},
			focus: function(){
				if(!$select.hasClass('ui-state-disabled')){
					$(this).addClass('ui-state-focus');
				}
			},
			blur: function(){
				$(this).removeClass('ui-state-focus');
			}
		});
		
		// bind custom events to the options div
		$options.bind({
			'close': function(e, others){
				others = others || false;
			
				// hides all other options but the one clicked
				if(others === true){
					$('div.ui-multiselect-options')
					.filter(':visible')
					.fadeOut(o.fadeSpeed)
					.prev('a.ui-multiselect')
					.removeClass('ui-state-active')
					.trigger('mouseout');
					
				// hides the clicked options
				} else {
					$select.removeClass('ui-state-active').trigger('mouseout');
					$options.fadeOut(o.fadeSpeed);
				}
			},
			'open': function(e, closeOthers){
				
				// bail if this widget is disabled
				if($select.hasClass('ui-state-disabled')){
					return;
				}
				
				// use position() if inside ui-widget-content, because offset() won't cut it.
				var offset = $select[ $select.closest('div.ui-widget-content').length ? 'position' : 'offset'](), 
					$container = $options.find('ul:last'), 
					top, width;
				
				// calling select is active
				$select.addClass('ui-state-active');
				
				// hide all other options
				if(closeOthers || typeof closeOthers === 'undefined'){
					$options.trigger('close', [true]);
				}
				
				// calculate positioning
				if(o.position === 'middle'){
					top = ( offset.top+($select.height()/2)-($options.outerHeight()/2) );
				} else if(o.position === 'top'){
					top = (offset.top-$options.outerHeight());
				} else {
					top = (offset.top+$select.outerHeight());
				}
				
				// calculate the width of the options menu
				width = $select.width()-parseInt($options.css('padding-left'),10)-parseInt($options.css('padding-right'),10);
				
				// select the first option
				$labels.filter('label:first').trigger('mouseenter').trigger('focus');
				
				// show the options div + position it
				$options.css({ 
					position: 'absolute',
					top: top+'px',
					left: offset.left+'px',
					width: width+'px'
				}).show();
				
				// set the scroll of the checkbox container
				$container.scrollTop(0);
				
				// set the height of the checkbox container
				if(o.maxHeight){
					$container.css('height', o.maxHeight);
				}
				
				o.onOpen.call($options[0]);
			},
			'toggle': function(){
				$options.trigger( $(this).is(':hidden') ? 'open' : 'close' );
			},
			'traverse': function(e, start, keycode){
				var $start = $(start), 
					moveToLast = (keycode === 38 || keycode === 37) ? true : false,
					
					// select the first li that isn't an optgroup label / disabled
					$next = $start.parent()[moveToLast ? 'prevAll' : 'nextAll']('li:not(.ui-multiselect-disabled, .ui-multiselect-optgroup-label)')[ moveToLast ? 'last' : 'first']();

				// if at the first/last element
				if(!$next.length){
					var $container = $options.find("ul:last");
					
					// move to the first/last
					$options.find('label')[ moveToLast ? 'last' : 'first' ]().trigger('mouseover');
					
					// set scroll position
					$container.scrollTop( moveToLast ? $container.height() : 0 );
					
				} else {
					$next.find('label').trigger('mouseenter');
				}
			},
			'toggleChecked': function(e, flag, group){
				var $inputs = (group && group.length) ? group : $labels.find('input');
				$inputs.not(':disabled').attr('checked', (flag ? 'checked' : '')); 
				updateSelected();
			}
		})
		.find('li.ui-multiselect-optgroup-label a')
		.click(function(e){
			// optgroup label toggle support
			var $checkboxes = $(this).parent().nextUntil('li.ui-multiselect-optgroup-label').find('input');
			
			$options.trigger('toggleChecked', [ ($checkboxes.filter(':checked').length === $checkboxes.length) ? false : true, $checkboxes]);
			o.onOptgroupToggle.call(this, $checkboxes.get());
			e.preventDefault();
		});
		
		// labels/checkbox events
		$labels.bind({
			mouseenter: function(){
				$labels.removeClass('ui-state-hover');
				$(this).addClass('ui-state-hover').find('input').focus();
			},
			click: function(e,caller){
				// if the label was clicked, trigger the click event on the checkbox.  IE6 fix
				e.preventDefault();
				$(this).find('input').trigger('click', [true]); 
			},
			keyup: function(e){
				switch(e.keyCode){
					case 27: // esc
						$options.trigger('close');
						break;
			
					case 38: // up
					case 40: // down
					case 37: // left
					case 39: // right
						$options.trigger('traverse', [this, e.keyCode]);
						break;
				
					case 13: // enter
						e.preventDefault();
						$(this).click();
						break;
				}
			}
		})
		.find('input')
		.bind('click', function(e, label){
			label = label || false;
			
			// stop this click from bubbling up to the label
			e.stopPropagation();

			// if the click originated from the label, stop the click event and manually toggle the checked state
			if(label){
				e.preventDefault();
				this.checked = this.checked ? false : true;
			}
		
			o.onCheck.call(this);
			updateSelected();
		});

		// remove the original input element
		$original.remove();
		
		// apply bgiframe if available
		if($.fn.bgiframe){
			$options.bgiframe();
		}
		
		// open by default?
		if(o.state === 'open'){
			$options.trigger('open', [false]);
		}
		
		// update the number of selected elements when the page initially loads, and use that as the defaultValue.  necessary for form resets when options are pre-selected.
		$select.find('input')[0].defaultValue = updateSelected();
	};
	
	// close each select when clicking on any other element/anywhere else on the page
	$(document).bind('click', function(e){
		var $target = $(e.target);

		if(!$target.closest('div.ui-multiselect-options').length && !$target.parent().hasClass('ui-multiselect')){
			$('div.ui-multiselect-options').trigger('close', [true]);
		}
	});

	// default options
	$.fn.multiSelect.defaults = {
		showHeader: true,
		maxHeight: 175, /* max height of the checkbox container (scroll) in pixels */
		minWidth: 200, /* min width of the entire widget in pixels. setting to 'auto' will disable */
		checkAllText: 'Check all',
		unCheckAllText: 'Uncheck all',
		noneSelectedText: 'Select options',
		selectedText: '# selected',
		selectedList: 0,
		position: 'bottom', /* top|middle|bottom */
		shadow: false,
		fadeSpeed: 200,
		disabled: false,
		state: 'closed',
		onCheck: function(){}, /* when an individual checkbox is clicked */
		onOpen: function(){}, /* when the select menu is opened */
		onCheckAll: function(){}, /* when the check all link is clicked */
		onUncheckAll: function(){}, /* when the uncheck all link is clicked */
		onOptgroupToggle: function(){} /* when the optgroup heading is clicked */
	};

})(jQuery);
